File: dart/Q31_submatrix_count/bin/solution3.dart
import 'dart:collection';

class Solution {
  int numSubmatrixSumTarget(List<List<int>> matrix, int target) {
    int count = 0;
    int rows = matrix.length;
    int cols = matrix[0].length;

    List<List<int>> sum_matrix = get_sums(matrix);

    for (int top = 0; top < rows; top++) {
      for (int bottom = top; bottom < rows; bottom++) {
        Map<int, int> map = HashMap();
        map[0] = 1;

        for (int c = 0; c < cols; c++) {
          int sum = sum_matrix[bottom][c];
          if (top > 0) {
            sum -= sum_matrix[top - 1][c];
          }

          count += map[sum - target] ?? 0;
          map[sum] = map.containsKey(sum) ? map[sum]! + 1 : 1;
        }
      }
    }

    return count;
  }

  List<List<int>> get_sums(List<List<int>> matrix) {
    int rows = matrix.length;
    int cols = matrix[0].length;
    List<List<int>> sum_matrix =
        List.generate(rows, (_) => List.filled(cols, 0));

    for (int r = 0; r < rows; r++) {
      for (int c = 0; c < cols; c++) {
        int top_sum = r == 0 ? 0 : sum_matrix[r - 1][c];
        int left_sum = c == 0 ? 0 : sum_matrix[r][c - 1];
        int intersection_sum = r > 0 && c > 0 ? sum_matrix[r - 1][c - 1] : 0;
        sum_matrix[r][c] = top_sum + left_sum - intersection_sum + matrix[r][c];
      }
    }
    return sum_matrix;
  }
}

File: main/Q23_matrix_search/solution3.go
package main

func searchMatrix3(matrix [][]int, target int) bool {
	return searchBox(matrix, target, 0, len(matrix[0])-1, len(matrix)-1, 0)
}

func searchBox(matrix [][]int, target int, top int, right int, bottom int, left int) bool {

	if top > bottom || left > right || target < matrix[top][left] || target > matrix[bottom][right] {
		return false
	}

	middle_col := left + (right-left)/2
	current_row := top

	for current_row <= bottom && matrix[current_row][middle_col] <= target {
		if matrix[current_row][middle_col] == target {
			return true
		}
		current_row++
	}

	return searchBox(matrix, target, current_row, middle_col-1, bottom, left) ||
		searchBox(matrix, target, top, right, current_row-1, middle_col+1)
}
